### PR TITLE
feat: Cancellative semigroups

### PR DESCRIPTION
**Pull request by YaelDillies**
Define semigroups that are cancellative on both sides.

This might sound a bit niche but I actually have a proof of the Cauchy-Davenport theorem that works for cancellative semigroups.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Review by eric-wieser — DISMISSED** [review]
Quoting `HierarchyDesign`:


> ## Instances
> 
> When adding a new typeclass `Z` to the algebraic hierarchy
> one should attempt to add the following constructions and results,
> when applicable:
> 
> * Instances transferred elementwise to products, like `Prod.Monoid`.
>   See `Mathlib.Algebra.Group.Prod` for more examples.
>   ```
>   instance Prod.Z [Z M] [Z N] : Z (M × N) := ...
>   ```
> * Instances transferred elementwise to pi types, like `Pi.Monoid`.
>   See `Mathlib.Algebra.Group.Pi` for more examples.
>   ```
>   instance Pi.Z [∀ i, Z $ f i] : Z (Π i : I, f i) := ...
>   ```
> * Instances transferred to `MulOpposite M`, like `MulOpposite.Monoid`.
>   See `Mathlib.Algebra.Opposites` for more examples.
>   ```
>   instance MulOpposite.Z [Z M] : Z (MulOpposite M) := ...
>   ```
> * Instances transferred to `ULift M`, like `ULift.Monoid`.
>   See `Mathlib.Algebra.Group.ULift` for more examples.
>   ```
>   instance ULift.Z [Z M] : Z (ULift M) := ...
>   ```
> * Definitions for transferring the proof fields of instances along
>   injective or surjective functions that agree on the data fields,
>   like `Function.Injective.monoid` and `Function.Surjective.monoid`.
>   We make these definitions `@[reducible]`, see note [reducible non-instances].
>   See `Mathlib.Algebra.Group.InjSurj` for more examples.
>   ```
>   @[reducible]
>   def Function.Injective.Z [Z M₂] (f : M₁ → M₂) (hf : f.Injective)
>     (one : f 1 = 1) (mul : ∀ x y, f (x * y) = f x * f y) : Z M₁ := ...
> 
>   @[reducible]
>   def Function.Surjective.Z [Z M₁] (f : M₁ → M₂) (hf : f.Surjective)
>     (one : f 1 = 1) (mul : ∀ x y, f (x * y) = f x * f y) : Z M₂ := ...
>   ```
> * Instances transferred elementwise to `Finsupp`s, like `Finsupp.semigroup`.
>   See `Mathlib.Data.Finsupp.Pointwise` for more examples.
>   ```
>   instance FinSupp.Z [Z β] : Z (α →₀ β) := ...
>   ```
> * Instances transferred elementwise to `Set`s, like `Set.monoid`.
>   See `Mathlib.Algebra.Pointwise` for more examples.
>   ```
>   instance Set.Z [Z α] : Z (Set α) := ...
>   ```
> * Definitions for transferring the entire structure across an equivalence, like `Equiv.monoid`.
>   See `Mathlib.Data.Equiv.TransferInstance` for more examples. See also the `transport` tactic.
>   ```
>   def Equiv.Z (e : α ≃ β) [Z β] : Z α := ...
>   /- When there is a new notion of `Z`-equiv: -/
>   def Equiv.ZEquiv (e : α ≃ β) [Z β] : by { letI := Equiv.Z e, exact α ≃Z β } := ...
>   ```

Right now this PR arguably creates more API holes than it fills!

**Comment by eric-wieser** [comment]
A good compromise would be "add an instance of your new two-sided cancellation typeclass for every instance of the one-sided ones"

File: Mathlib/Data/Fin/Basic.lean [hunk 1]
```diff
@@ -1191,20 +1191,12 @@ instance instInvolutiveNeg (n : ℕ) : InvolutiveNeg (Fin n) where
 #align fin.involutive_neg Fin.instInvolutiveNeg
 
 /-- Note this is more general than `Fin.addCommGroup` as it applies (vacuously) to `Fin 0` too. -/
-instance instIsCancelAdd (n : ℕ) : IsCancelAdd (Fin n) where
```
Comment (eric-wieser) [inline Mathlib/Data/Fin/Basic.lean:1194]: What's the link between `IsCancelAdd` and your new typeclass? Do we have the instance that one implies the other?

File: Mathlib/Combinatorics/SimpleGraph/mwe.lean
Comment (eric-wieser) [inline Mathlib/Combinatorics/SimpleGraph/mwe.lean:1]: I assume this was an accident

File: Mathlib/Algebra/Group/Defs.lean [hunk 2]
```diff
@@ -752,14 +771,8 @@ instance (priority := 100) CancelCommMonoid.toCancelMonoid (M : Type u) [CancelC
...
-/-- Any `CancelMonoid G` satisfies `IsCancelMul G`. -/
-@[to_additive toIsCancelAdd "Any `AddCancelMonoid G` satisfies `IsCancelAdd G`."]
-instance (priority := 100) CancelMonoid.toIsCancelMul (M : Type u) [CancelMonoid M] :
-    IsCancelMul M :=
-  { mul_left_cancel := LeftCancelSemigroup.mul_left_cancel
-    mul_right_cancel := RightCancelSemigroup.mul_right_cancel }
-#align cancel_monoid.to_is_cancel_mul CancelMonoid.toIsCancelMul
-#align add_cancel_monoid.to_is_cancel_add AddCancelMonoid.toIsCancelAdd
+#noalign cancel_monoid.to_is_cancel_mul
+#noalign add_cancel_monoid.to_is_cancel_add
```
Comment (eric-wieser) [inline Mathlib/Algebra/Group/Defs.lean:775]: I think these should be aligned with the semigroup versions, to the caller (using only explicit arguments) they are equivalent.

File: Mathlib/Algebra/Group/OrderSynonym.lean [hunk 3]
```diff
@@ -56,6 +56,9 @@ instance [h : LeftCancelSemigroup α] : LeftCancelSemigroup αᵒᵈ := h
 @[to_additive]
 instance [h : RightCancelSemigroup α] : RightCancelSemigroup αᵒᵈ := h
 
+@[to_additive]
+instance [h : CancelSemigroup α] : CancelSemigroup αᵒᵈ := h
```
Comment (eric-wieser) [inline Mathlib/Algebra/Group/OrderSynonym.lean:60]: Can you add this for `Lex` too in the same file?

File: Mathlib/Algebra/Group/Defs.lean [hunk 4]
```diff
@@ -426,6 +426,24 @@ instance (priority := 100) RightCancelSemigroup.toIsRightCancelMul (G : Type u)
...
+class CancelSemigroup (α : Type*) extends LeftCancelSemigroup α, RightCancelSemigroup α
+
+attribute [to_additive existing] CancelSemigroup.toRightCancelSemigroup
+
+/-- Any cancellative semigroup has cancellative multiplication. -/
+@[to_additive toIsCancelAdd "Any cancellative semigroup has cancellative addition."]
+instance (priority := 100) CancelSemigroup.toIsCancelMul (M : Type*) [CancelSemigroup M] :
+    IsCancelMul M :=
+  { mul_left_cancel := LeftCancelSemigroup.mul_left_cancel
+    mul_right_cancel := RightCancelSemigroup.mul_right_cancel }
```
Comment (eric-wieser) [inline Mathlib/Algebra/Group/Defs.lean:445]: Does this work?
```suggestion
  { }
```

(or perhaps a lone `where`)

It should be implied by the instance above.

**Comment by eric-wieser** [comment]
This generally looks good, but I'm a little worried that `CancelSemigroup` is solving a problem that we could already solve with `IsCancelMul`; did you deliberately choose to avoid the latter, or were you just not aware of it?

We could consider going in the other direction and delete `(Left|Right|)CancelMonoid`, using `IsCancelAdd` everywhere.

**Comment by YaelDillies** [comment]
My proof of Cauchy-Davenport requires left and right cancellativity and associativity. I don't need a unit element.
* `LeftCancelMonoid`/`RightCancelMonoid` provide me one of the cancellativities and associativity.
* `IsCancelMul` provides me the cancellativities but not the associativity.
* `CancelMonoid` gives me a unit element I don't need.

**Comment by eric-wieser** [comment]
`IsCancelMul` is a mixin, so you can write `[Semigroup M] [IsCancelMul M]`

**Comment by YaelDillies** [comment]
Sure, but at any rate it's weird that we have left and right cancellativity and not the combination of both, especially because we do so for monoids.

**Comment by urkud** [comment]
When `Is*CancelMul` was added, the plan was to drop the `*CancelMonoid` classes and use the new mixins as needed. Nobody implemented this plan but this PR goes in the opposite direction.

**Comment by eric-wieser** [comment]
Indeed, that was discussed in https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/Preferred.20spelling.20of.20cancellative.20semigroups/near/402284870

**Comment by YaelDillies** [comment]
Closing since the lemmas I needed have been generalised to the mixins.

**Pull request closed without merging**